sunpy/map/mapbase.py

```diff
@@ -546,6 +546,7 @@ def wcs(self):
         # array with sequence error.
         w2.wcs.cdelt = u.Quantity(self.scale)
         w2.wcs.crval = u.Quantity([self._reference_longitude, self._reference_latitude])
+        w2.wcs.lonpole = 180. if w2.wcs.crval[1] < w2.wcs.theta0 else 0.
```

Comment:
Why are we overwriting the value if it's given in the header here? And doesn't `wcslib` take care of setting the appropriate default value if it isn't given in the header?

Reply:
Well, hmm.  Presumably the reason this entire code block exists is that we don't want to create the WCS from the header values directly.  The WCS is created above *without* using header values, and then all of these lines set WCS values based on our properties.  If you manually change the reference coordinate, you have to manually update `lonpole` to be the appropriate default value.

To fully follow this model, I can create a `GenericMap._lonpole` (private) property that would use the `lonpole` keyword if it is in the header, otherwise return the appropriate default.

Reply:
To expand on my comment, it seems like `wcslib` automatically sets LONPOLE if it's not in the header. So this might be an upstream `wcslib` issue, or an issue with a particular data source. Do you have a self contained example that motivated this PR?
```python
from sunpy.data.sample import AIA_193_IMAGE
from sunpy.map import Map

m = Map(AIA_193_IMAGE)
m.meta.pop('lonpole', None)
print('lonpole' in m.meta)  # False
print(m.wcs.to_header()['LONPOLE'])  # 180
```

Reply:
This is an issue with how we construct the WCS.

The following line creates the WCS instance without using many of the header values:
https://github.com/sunpy/sunpy/blob/8d24c7ae268dfcb6d05f52193847333e9a166b5a/sunpy/map/mapbase.py#L526
The created WCS instance has `LONPOLE` set to its appropriate default value for the _default_ value of the reference coordinate (e.g., (0, 0) in Stonyhurst heliographic coordinates).

Then, the following line (potentially) modifies the reference coordinate in the WCS instance to something other than the default reference coordinate (for example, to one with a negative latitude in Stonyhurst heliographic coordinates):
https://github.com/sunpy/sunpy/blob/8d24c7ae268dfcb6d05f52193847333e9a166b5a/sunpy/map/mapbase.py#L548

However, the value for `LONPOLE` has _already been set_ in the WCS instance, and so now you're in the situation where `LONPOLE` may no longer be at its appropriate default value for the non-default reference coordinate.  Note that the user may have intentionally set `LONPOLE` to a non-default value, so I don't fault `wcslib` for not automatically updating `LONPOLE` when `CRVAL` is updated.

Reply:
This is the easiest way to show this.  Again, I don't see this as a bug in `wcslib`.

```python
>>> import astropy.units as u
>>> from astropy.coordinates import SkyCoord
>>> from astropy.wcs import WCS
>>> import sunpy.map

>>> header1 = sunpy.map.make_fitswcs_header(
...     (180, 360),
...     SkyCoord(0*u.deg, 10*u.deg, frame='heliographic_stonyhurst', obstime='2021-07-21'),
...     scale=(1, 1)*u.deg/u.pix,
...     projection_code='CAR'
... )
>>> del header1['lonpole']  # before PR #5448, this value may be incorrect
>>> wcs1 = WCS(header1)
>>> print(wcs1.wcs.crval)
>>> print(wcs1.wcs.lonpole)  # default value is 0 for HGS reference latitude >= 0
[ 0. 10.]
0.0

>>> header2 = sunpy.map.make_fitswcs_header(
...     (180, 360),
...     SkyCoord(0*u.deg, -10*u.deg, frame='heliographic_stonyhurst', obstime='2021-07-21'),
...     scale=(1, 1)*u.deg/u.pix,
...     projection_code='CAR'
... )
>>> del header2['lonpole']  # before PR #5448, this value may be incorrect
>>> wcs2 = WCS(header2)
>>> print(wcs2.wcs.crval)
>>> print(wcs2.wcs.lonpole)  # default value is 180 for HGS reference latitude < 0
[  0. -10.]
180.0

>>> wcs1a = wcs1.copy()
>>> wcs1a.wcs.crval = wcs2.wcs.crval  # modify the reference latitude to be < 0
>>> print(wcs1a.wcs.crval)
>>> print(wcs1a.wcs.lonpole)  # LONPOLE is still at 0 instead of being changed to 180
[  0. -10.]
0.0
```

Reply:
I agree re. the fixes to `make_fitswcs_header`. I don't understand
> The following line creates the WCS instance without using many of the header values:
> https://github.com/sunpy/sunpy/blob/8d24c7ae268dfcb6d05f52193847333e9a166b5a/sunpy/map/mapbase.py#L526

This uses all the information in the FITS header. We then overwrite some of that information using the map properties, but if there is a reference coordinate in the FITS header it is set in the line above, and `wcslib` proceeds to set the correct default `LONPOLE` value. So I'm still not convinced that a change needs to be made in `GenericMap.wcs`. To illustrate this, on the current main branch:
```python
import astropy.units as u
from astropy.coordinates import SkyCoord
from astropy.wcs import WCS
import sunpy.map
import numpy as np

shape = (180, 360)
header1 = sunpy.map.make_fitswcs_header(
    (180, 360),
    SkyCoord(0*u.deg, -10*u.deg, frame='heliographic_stonyhurst', obstime='2021-07-21'),
    scale=(1, 1)*u.deg/u.pix,
    projection_code='CAR'
)
print(header1['lonpole'])  # 0 (incorrect, but fixed by #5448)
del header1['lonpole']
wcs1 = WCS(header1)
print(wcs1.wcs.lonpole)  # 180 (correct)

data = np.ones(shape)
m = sunpy.map.Map(data, header1)
print(m.wcs.wcs.lonpole)  # 180 (correct)
```

Reply:
Hmm, okay, that's convincing.  I suspect that I must have been testing with a header with a _wrong_ value for `LONPOLE`, as opposed to a missing value, and I got confused and misinterpreted the behavior of this code block.  I'll take out this line.  Thanks!